Parse `impl` blocks, `type` and `const` declarations

Bromeon opened the pull request:
This pull request takes #27 as a base.

Supports parsing of two impl block schemes:
* Inherent: `impl Type { ... }`
* Trait: `impl Trait for Type { ... }`

Each block can contain members of three different categories:
* functions
* constants
* associated types

Examples:
```rust
// Inherent impl
impl<'a, T: Clone, const N: i8> MyStruct<'a, T, N> {}

// Trait impl
impl MyTrait for MyStruct {
    pub type MyType = std::string::String;

    fn new(i: i32, b: bool) -> Self {
        Self {}
    }

    #[attr]
    const fn set_value(&mut self, s: String) {}

    const CONSTANT: i8 = 24 + 7;
}
```

---

There are a few open questions on my side:
1. I plan to also add support for `trait`s. Regarding some of the new API types I added, there's the option to reuse them (forcing users to unwrap options) or to make dedicated, type-safe separate ones for `trait` and `impl` contexts. What is your preference? This would also affect the naming (e.g. I'm not happy with `TypeDefinition`). 
  Examples:
    * Trait: `const CONSTANT: i8;`
      Impl: `const CONSTANT: i8 = value;`
      either we have 1 type representing both, with optional `=` and `value`, or 2 separate types.
    * Trait: `type AssocTy: Clone;` -- can have bounds, but (at the moment) no default type initializers
      Impl: `type AssocTy = MyType;` -- must have type initializers, no bounds

1. I was thinking to add `GenericArg`/`GenericArgList` to represent `<'a, T, U, N>` -- in addition to `GenericParam/GenericParamList` for `<'a: 'static, T, U: Clone, const N: usize>`. Would that make sense?
   If yes, we could probably remove `InlineGenericArgs` and directly allow to convert from parameters (unless you say we can't afford that copy).

1. Is there a reason why the `Debug` impl for `GenericParam` does not include the `tk_prefix`? Might be good to check if it's a lifetime, a const or a type.
   https://github.com/PoignardAzur/venial/blob/b490dd70226cb2b46d63849154bd06a35ee566d3/src/types.rs#L486-L493

1. There might be some room to make the naming more consistent (e.g. `FunctionParameter` vs. `GenericParam`) as well as other smaller refactorings, but it probably makes more sense to work out a proposal to parse `trait` first 🙂

Bromeon commented:
I realized that the previous structures couldn't represent some syntax; concretely:
* `Ident` is not enough for the _Trait_ and _Self_ parts in `impl Trait for Self`.
  These can be arbitrarily long type paths, or even non-ident things such as `()`.
* `GenericParam` cannot support anything beyond simple idents, while generic _arguments_ can be arbitrarily complex types.

So I just used `TyExpr`. Since that just leaves verbatim tokens, I added a convenience function `TyExpr::as_path()` that can (partially) parse `path::to::Type<other::Arg, 32, 'a>`. We don't yet have a type representing generic arguments, so the argument list is left verbatim. 

After some feedback I can also clean up the commits a bit.

PoignardAzur commented on src/tests.rs:
```diff
@@ -988,3 +986,109 @@ fn add_where_item() {
+    };
+
+    let impl_decl = parse_declaration_checked(expr);
+    assert_debug_snapshot!(impl_decl);
```
Have you considered using `assert_quote_snapshot` ? (defined higher in the file)

PoignardAzur commented on src/types.rs:
```diff
@@ -335,13 +457,64 @@ pub struct WhereClauseItem {
     pub tokens: Vec<TokenTree>,
 }
 
+impl TyExpr {
+    /// Tries to parse this type in the common form `path::to::Type<'a, other::Type>`.
+    ///
+    /// The 1st element of the tuple is the path, e.g. `path::to::Type`.
+    /// The 2nd element are the raw tokens _inside_ the generic argument list, if available, e.g. `'a, other::Type`.
+    ///
+    /// If this type has any other form, `None` is returned.
+    pub fn as_path(&self) -> Option<(Vec<Ident>, Option<&[TokenTree]>)> {
+        let found_split = self
+            .tokens
+            .iter()
+            .enumerate()
+            .find(|(_, tk)| matches!(tk, TokenTree::Punct(punct) if punct.as_char() == '<'));
+
+        let (prefix, generic_args) = if let Some((split_pos, _)) = found_split {
+            let (prefix, generic_args) = self.tokens.split_at(split_pos);
+
+            let closing_angle_bracket = generic_args.last();
+            if matches!(closing_angle_bracket, Some(TokenTree::Punct(punct)) if punct.as_char() == '>')
+            {
+                (prefix, Some(&generic_args[1..generic_args.len() - 1]))
+            } else {
+                (prefix, None)
+            }
+        } else {
+            // could be 'path::to::Type' without generic args, or something entirely different
+            (self.tokens.as_slice(), None)
+        };
+
+        let path_tokens = tokens_from_slice(prefix);
+        if let Some(path_elems) = try_consume_path(path_tokens) {
+            return Some((path_elems, generic_args));
+        }
+
+        None
+    }
+}
```
This whole block should be in `types_edition.rs`.

PoignardAzur replied:
Also, `as_path` isn't a terribly descriptive method name. (though I don't have a better name in mind)

Bromeon replied:
Renamed to `interpret_as_path`.

"Path" may sound generic, but [it's the official naming](https://doc.rust-lang.org/reference/paths.html) for token trees like `path::to::Type<'a, other::Type>`.

PoignardAzur replied:
The "as_xxx" naming matches the rest of the crate, though I guess it's not idiomatic rust. My problem is with "path".

> "Path" may sound generic, but [it's the official naming](https://doc.rust-lang.org/reference/paths.html) for token trees like `path::to::Type<'a, other::Type>`.

Yes, but the definition of those paths in the doc doesn't match yours. Eg `mycrate::Foo<Bar>::AssociatedType` is a valid path, though in practice this syntax is rarely used.

Depending on the use case you want for this utility function, this is something to consider.

Bromeon replied:
Good point. The general question is also, how far should the parsing go? Path expressions can become arbitrarily complex, e.g.
```rs

PoignardAzur commented on src/types.rs:
```diff
@@ -133,6 +135,91 @@ pub struct Union {
+    pub body: ImplBody,
+}
+
+/// The `{ }` group representing the body of an `impl` block.
+#[derive(Clone, Debug)]
+pub struct ImplBody {
+    pub members: Vec<ImplMember>,
+    pub tk_braces: GroupSpan,
+}
```
Could probably be inlined in `Impl`? AFAICT `Impl` only ever stores one `ImplBody`, never zero or multiple.

PoignardAzur commented on src/parse_impl.rs:
```diff
@@ -0,0 +1,173 @@
+                    ImplMember::AssocTy(assoc_ty)
+                }
+                "const" => {
+                    // needs 2-lookahead to differentiate `const IDENT` and `const fn`
```
I don't like this hack.

If nothing else, it doesn't handle the case where you have `const async fn foobar()`.

I think it would be better to first call `consume_fn_qualifiers`, then check for the fn keyword, and if you can't find it, "fall back" to the const case. Extract the const qualifier, throw the other qualifiers away, keep going. (Though it would give ugly code too.)

PoignardAzur commented on src/tests.rs:
```diff
@@ -988,3 +986,109 @@ fn add_where_item() {
     assert_quote_snapshot!(type_with_args);
 }
+
+// =================
```
Looks like you went the extra mile to consider multiple cases for the impl parser. I definitely appreciate it!

PoignardAzur commented:
> I plan to also add support for traits. Regarding some of the new API types I added, there's the option to reuse them (forcing users to unwrap options)

I'd lean towards re-using them.

Note that venial aims to cover anything that is syntactically valid, even if it's not semantically valid. For instance, this compiles:

```rust
#[cfg(FALSE)]
const X: i32;
```

> I was thinking to add GenericArg/GenericArgList to represent <'a, T, U, N> -- in addition to GenericParam/GenericParamList for <'a: 'static, T, U: Clone, const N: usize>. Would that make sense?
If yes, we could probably remove InlineGenericArgs and directly allow to convert from parameters (unless you say we can't afford that copy).

I guess? I haven't really considered it.

I think we'd want names that differentiate between the two less ambiguously. I like `InlineGenericArgs`, but I'm open to suggestions.

> Is there a reason why the Debug impl for GenericParam does not include the tk_prefix? Might be good to check if it's a lifetime, a const or a type.

I don't remember. Probably not.

Bromeon commented:
@PoignardAzur Sorry for the delay. In my initial implementation, a lot of stuff was "half-baked". By doing things a bit more properly, I realized that several things were missing and took the chance to add quite a bit of functionality in the process, while also addressing your feedback 😬 I apologize if this PR got a bit too big.

New Features in this PR:
* Top-level `const`, `type` and `impl` declarations
* `const`, `type` and `fn` declarations inside `impl` blocks
* Path parsing, new types `Path` and `PathSegment`
   * `TyExpr::as_path()` lazily constructs a path from a type when needed
* Generic argument parsing, new types `GenericArg` + `GenericArgList`
   * `InlineGenericArgs::to_owned_args()` to convert to `GenericArgList`
* `ValueExpr` type that behaves like `TyExpr`
   * I'm not sure if this is needed. There are scenarios like generic arguments, where I cannot differentiate between the two, since venial doesn't do expression parsing. Should we merge them (under which name?) or just use `TyExpr` everywhere?

Also added tests for everything.

I left the commits split, so it's easier for you to follow the changes. I can squash them to only a handful if you prefer.

PoignardAzur commented:
Will take a look soon.

PoignardAzur commented:
FYI, I'm going to vacation on wednesday (well, kind of, but the gist is I'll have less time for open source) but I'd like to review this before.

Ping me tomorrow if I haven't gone back to you yet.

Bromeon commented:
@PoignardAzur the day is not over yet, but here's a ping, to not push it too far into the evening 😉

PoignardAzur commented on src/parse_fn.rs:
```diff
@@ -224,3 +145,103 @@ pub(crate) fn parse_fn_params(tokens: TokenStream) -> Punctuated<FunctionParamet
+/// Panics when the following tokens do not constitute a function definition, with one exception:
+/// when `const` is followed by an identifier which is not `fn`, then `None` is returned. This is to
+/// allow fallback to a constant declaration (both can begin with the `const` token).
+pub(crate) fn try_consume_fn(
```
This function should be called `consume_fn`.

Overall, the naming convention so far has been that `consume_xxx` functions in venial return None if the syntax item is absent, and panic if the *beginning* of the syntax item is present but the expect continuation isn't.

Eg: `consume_where_clause` will return None if it doesn't find a `where` token, but will panic if it gets `where "hello world"`.

Bromeon replied:
Done.

PoignardAzur commented on src/parse_impl.rs:
```diff
@@ -0,0 +1,167 @@
+
+type TokenIter = Peekable<proc_macro2::token_stream::IntoIter>;
+
+pub(crate) fn consume_for(tokens: &mut TokenIter) -> Option<Ident> {
+    if let Some(TokenTree::Ident(ident)) = tokens.peek() {
+        if ident == "for" {
+            let tk_for = ident.clone();
+            tokens.next();
+            return Some(tk_for);
+        }
+    }
+
+    None
+}
```
Nitpick: this should be a `match` to be consistent with how the rest of the project does it.

Bromeon replied:
Done.

PoignardAzur commented on src/parse_impl.rs:
```diff
@@ -0,0 +1,167 @@
+    }
+}
+
+pub(crate) fn parse_impl_members(token_group: Group) -> (Vec<ImplMember>, GroupSpan) {
```
This misses inner attributes, eg:

```rust
impl Foobar {
    #![hello]
}
```

Bromeon replied:
Implemented.

Also renamed `tk_hashbang` to `tk_hash`, which is more correct.

PoignardAzur commented on src/parse_type.rs:
```diff
@@ -109,6 +110,115 @@ pub(crate) fn consume_generic_params(tokens: &mut TokenIter) -> Option<GenericPa
     })
 }
 
+fn consume_generic_arg(tokens: Vec<TokenTree>) -> GenericArg {
```
To follow venial's naming convention, this should be called `parse_generic_arg`.

Bromeon replied:
Done.

PoignardAzur commented on src/parse_type.rs:
```diff
@@ -109,6 +110,115 @@ pub(crate) fn consume_generic_params(tokens: &mut TokenIter) -> Option<GenericPa
+        }
+    }
+
+    // Then, try parsing Item = ...
+    // (there is at least 1 token, so unwrap is safe)
+    let before_ident = tokens.clone();
+    if let TokenTree::Ident(ident) = tokens.next().unwrap() {
+        if let Some(TokenTree::Punct(punct)) = tokens.next() {
+            if punct.as_char() == '=' {
+                let remaining: Vec<TokenTree> = tokens.collect();
+
+                return GenericArg::Binding {
+                    ident,
+                    tk_equals: punct,
+                    ty: TyExpr { tokens: remaining },
+                };
+            }
+        }
+    }
```
This doesn't cover generic bindings (eg `LendingIterator<Item<'_> = XXX>`), though it doesn't really matter for now.

Bromeon replied:
Added `TODO`

PoignardAzur commented on src/parse_utils.rs:
```diff
@@ -170,3 +176,64 @@ pub(crate) fn consume_comma(tokens: &mut TokenIter) -> Option<Punct> {
+/// Parse `::`, as in path separator or turbofish.
+///
+/// Does not advance `tokens` if the double colon is not found.
+pub(crate) fn try_consume_colon2(tokens: &mut TokenIter) -> Option<[Punct; 2]> {
```
Should be `consume_colon2` (or consume_double_colon?)

Bromeon replied:
Done.

PoignardAzur commented on src/parse_utils.rs:
```diff
@@ -170,3 +176,64 @@ pub(crate) fn consume_comma(tokens: &mut TokenIter) -> Option<Punct> {
+            break;
+        }
+
+        println!("-----");
+        dbg!(&segments);
+        dbg!(&tokens);
```
Leftover debug statements.

Bromeon replied:
Removed.

PoignardAzur commented on src/parse_utils.rs:
```diff
@@ -170,3 +176,64 @@ pub(crate) fn consume_comma(tokens: &mut TokenIter) -> Option<Punct> {
+        }
+    }
+
+    *tokens = before;
```
I'm not crazy about this part.

I get why it's necessary, but it has the potential to clone a lot of token, which is annoying given that the whole point of venial is to give good performance.

On the long term, I'd rather have a token iterator with more lookahead for that kind of situation, though for now we'll probably merge this as-is.

Bromeon replied:
Added `TODO`

PoignardAzur commented on src/tests.rs:
```diff
@@ -778,6 +825,26 @@ fn parse_fn_prototype() {
 }
 
+#[test]
+fn parse_fn_mut_param() {
+    let func = parse_declaration(quote! {
+        fn prototype(a: i32, mut b: f32) -> String;
+    })
+    .unwrap();
+
+    assert_debug_snapshot!(func);
+}
```
Minor nitpick: I'd rather these tests be a bit more consistent with existing tests:

- Use assert_quote_snapshot
- Use parse_declaration_checked
- Use `quote!` with parentheses
- Add a consume_generic_args_checked function that does the same thing as parse_declaration_checked

Though none of these are blockers, it's more for the code aesthetic ^^

Bromeon replied:
Done, except `assert_quote_snapshot`, here I wasn't sure what exactly you meant.

This is already part of `parse_declaration_checked`, no? `assert_debug_snapshot` seems more meaningful for the "data model" representation (and not the token one). Unless this is very important, I'd prefer if this one could be postponed, otherwise this PR is never-ending 🙂

PoignardAzur commented on src/parse_impl.rs:
```diff
@@ -0,0 +1,167 @@
+        true,
+    );
+
+    let tk_equals = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == '=' => punct,
+        _ => panic!("cannot parse constant"),
+    };
+
+    let value_tokens = consume_stuff_until(
+        tokens,
+        |tt| matches!(tt, TokenTree::Punct(punct) if punct.as_char() == ';'),
+        true,
+    );
```
Constants without `= xxx` are syntactically valid and should be accepted.

Bromeon replied:
Implemented, the `=` and value are now optional:
```rs
pub struct Constant {
    pub attributes: Vec<Attribute>,
    pub vis_marker: Option<VisMarker>,
    pub tk_const: Ident,
    pub name: Ident,
    pub tk_colon: Punct,
    pub ty: TyExpr,
    /// The '=' sign is optional; constants without initializer are syntactically valid.
    pub tk_equals: Option<Punct>,
    /// The initializer value is optional; constants without initializer are syntactically valid.
    pub initializer: Option<ValueExpr>,
    pub tk_semicolon: Punct,
}
```

PoignardAzur commented on src/parse_impl.rs:
```diff
@@ -0,0 +1,167 @@
+    }
+}
+
+pub(crate) fn consume_assoc_ty(
+    tokens: &mut TokenIter,
+    attributes: Vec<Attribute>,
+    vis_marker: Option<VisMarker>,
+) -> TyDefinition {
```
This should be called `consume_type_definition`. If called from the top level, it might parse a type alias, not an associated type.

(Also, for consistency, it should arguably return Option)

Bromeon replied:
Renamed and optionalized 🙂

PoignardAzur commented on src/parse_impl.rs:
```diff
@@ -0,0 +1,167 @@
+        _ => panic!("cannot parse associated type"),
+    };
+
+    let tk_equals = match tokens.next() {
+        Some(TokenTree::Punct(punct)) if punct.as_char() == '=' => punct,
+        _ => panic!("cannot parse associated type"),
+    };
```
This assumes that every type definition is of the form `type Foo = some::Stuff;`. But:

- Type definitions may have bounds (eg `type Foo: Clone;`)
- Type definitions may have generic parameters (eg `type Item<Foo> = Bar;`)
- Type definitions may have where-clauses (eg `type Foo where T: Debug = Bar;`)
- Type definitions are not required to have a type (eg `type Foo;`)

See the full definition there: https://doc.rust-lang.org/reference/items/type-aliases.html

Bromeon replied:
If possible, I'd like to defer this to a follow-up PR about traits, then we could have one `TyDefinition` which supports both.

PoignardAzur reviewed (approved):
Overall, this all seems like very high-quality code.

The biggest problems are the incomplete parsers for type and const declarations, but it's always better than no implementation at all, so I wouldn't mind merging it now.

Do you want me to give you write access to the repository? Since I don't intend to maintain the project in the coming months, you might want a tighter feedback loop. And while I made a lot of small nitpicks, overall the quality of your contributions seems high enough I don't need to double-check everything.

Bromeon commented:
> The biggest problems are the incomplete parsers for type and const declarations, but it's always better than no implementation at all, so I wouldn't mind merging it now.

I agree here, but maybe I should highlight missing parts (e.g. as `// TODO`) so they could potentially be added later (or at least they document the limitations). I can do this for some of the inline code comments, which require bigger changes and would further delay this PR.

> Do you want me to give you write access to the repository? Since I don't intend to maintain the project in the coming months, you might want a tighter feedback loop. And while I made a lot of small nitpicks, overall the quality of your contributions seems high enough I don't need to double-check everything.

Thanks a lot for the compliment and the trust, it's appreciated 🙂

Write access might be nice for quicker integration, thanks for offering. Even though, I have to say your feedback in PRs has also been very valuable; I still need to understand some of the conventions and practices of venial. I don't know if you plan on reviewing things after some months, but I guess you could also directly clean up the `main` branch whenever you see fit (for smaller nitpicks, it's often faster than expressing your thoughts into words, letting me implement them and reviewing them again).

Maybe more generally, would that be more of a hiatus for a couple of months, or are you generally not sure about long-term plans? Are there more people involved or potentially interested in venial's maintenance? With godot-rust and a couple other open-source projects, I could probably implement some things every now and then, but realistically I don't see how I can spend enough time to help review PRs etc. I do see a lot of potential for a lightweight parser though, and know some people who think similarly.

PoignardAzur commented:
> I agree here, but maybe I should highlight missing parts (e.g. as // TODO) so they could potentially be added later (or at least they document the limitations).

Yeah, that's good practice.

> I don't know if you plan on reviewing things after some months,

I still plan on reviewing incoming PRs (there's few enough it's not exactly a drag).

> but I guess you could also directly clean up the main branch whenever you see fit (for smaller nitpicks, it's often faster than expressing your thoughts into words, letting me implement them and reviewing them again).

That's the idea. :grin:

> Maybe more generally, would that be more of a hiatus for a couple of months, or are you generally not sure about long-term plans?

I don't really have long-term plans for venial. As I've said elsewhere, the initial benchmarks have been somewhat disappointing (it's faster than syn, but not so much that I'd expect mass adoption). I mostly see the project as an informative experiment.

My plan is to make a 1.0 release at some point this year, with an attached post-mortem, and then I'll only fix issues and merge PRs.

> Are there more people involved or potentially interested in venial's maintenance? With godot-rust and a couple other open-source projects, I could probably implement some things every now and then, but realistically I don't see how I can spend enough time to help review PRs etc. I do see a lot of potential for a lightweight parser though, and know some people who think similarly.

I'm not trying to dump the project in your lap ^^.

There aren't too many people involved in the project (as you can see from the commit history). If you know people who'd be interested in getting involved, sure, that would be helpful, but I don't really expect it.

Again, I don't plan on archiving the project. If people submit PRs, I'll review them. I just don't plan to have a very active role in the future.

(And honestly, given this was only side project I tried to do in a week to distract myself from Druid, I'm pretty impressed with how far it has come)

Bromeon commented:
> I'm not trying to dump the project in your lap ^^.

No worries, that's not how I understood it, just wanted to make sure you're not disappointed if my contributions decrease in frequency over time 😉 



> I don't really have long-term plans for venial. As I've said elsewhere, the initial benchmarks have been somewhat disappointing (it's faster than syn, but not so much that I'd expect mass adoption). I mostly see the project as an informative experiment. [...] (And honestly, given this was only side project I tried to do in a week to distract myself from Druid, I'm pretty impressed with how far it has come)

In godot-rust, syn is a major compile time offender (total compile time 206s fresh, of which syn is 18%):

![grafik](https://user-images.githubusercontent.com/708488/183638207-b06dc0a0-34b4-4209-9f6a-9b019e7c6902.png)


Now I won't rewrite that all to venial, so it's hard to tell how much difference there really is. But what I _am_ going to try is use venial for the next version of godot-rust (GDExtension binding) and see how far I can push it. I'm a huge fan of lightweight libraries -- could be that I won't reach my goals in the end, but I think venial deserves to get a chance in a larger project 🚀 

syn is an extremely powerful and versatile library, but two things bug me in particular:
1. Libraries tend to be overly liberal in using it and enabling many crate features, so a dependency tree often has almost all features enabled. This in turn means that even libraries that only use the very basic syn configuration pay for the complex parsing.
2. Most of the times, what's needed is parsing "headers" (attributes, struct fields, function signatures, etc) and not function bodies or expressions. syn however parses everything (see above point). I can imagine that the extra parsing time and all the intermediate allocations to store symbols can add up quite a bit over many crates.
 
A syn 2.0 could possibly address 1) by modularizing the crates, and 2) by providing lazy evaluation for rarely used items. But this is a mostly speculative idea without benchmarking. venial is a good trade-off here.

Bromeon commented:
I addressed your feedback:
* improved consistency in naming and tests
* support for `const C: i32;`
* support for `#![inner_attr]` in `impl` blocks
* added `TODOs` for limitations

The `TyDefinition` I could extend in another PR, which would add support for traits.

Another thought -- once venial supports a lot more cases than just basic structs and some functions, should we advertise it a bit (Twitter, Reddit etc.)? Maybe after a `0.5` release or so...

PoignardAzur commented:
> Another thought -- once venial supports a lot more cases than just basic structs and some functions, should we advertise it a bit (Twitter, Reddit etc.)? Maybe after a 0.5 release or so...

Yeah, that's definitely the plan!

Bromeon commented:
Do you think we can merge this PR as it is now? 🙂

PoignardAzur commented:
Done.

PoignardAzur commented:
That comment you added makes me think it might be worth adding an `ARCHITECTURE.md` file to the project.

Bromeon commented:
That sounds like a good idea! I haven't seen the `ARCHITECTURE.md` convention before, maybe `CONTRIBUTING.md` is more common as that's the file that contributors will likely read second (after the ReadMe)? Or a link to that file would also work.

PoignardAzur commented:
I haven't seen it in the wild too much (or at all?), I'm referring to the kind of documentation described in this post: https://matklad.github.io/2021/02/06/ARCHITECTURE.md.html

I see it as being a bit different from `CONTRIBUTING.md` (one is more of a checklist, the other is a high-level view of the project) but it's not that important.
